stdeb.cfg

```diff
@@ -1,4 +1,5 @@
 [colcon-ros]
 Depends3: python3-catkin-pkg-modules, python3-colcon-cmake, python3-colcon-core (>= 0.3.1), python3-colcon-pkg-config, python3-colcon-python-setup-py, python3-colcon-recursive-crawl
+Recommends3: python3-colcon-argcomplete
```

Comment:
If a user doesn't use a shell which is supported by `colcon-argcomplete` then this is an unnecessary dependency. A similar case would be if this package is only installed in a CI environment where completion is not used.

In my mind `colcon-common-extensions` was the package recommended to install if you wants a collections of "useful common" extensions. Otherwise the user should be able to install only the Debian packages needed (without anything being pulled in which they don't want).

Reply:
In this case, I propose to use Suggests3 ?
(idea is to provide a full tree dependancy for system admin)

Debian Package recommendation say :
https://www.debian.org/doc/manuals/debian-faq/ch-pkg_basics.en.html (chap 7.9)

-    Package A recommends Package B, if the package maintainer judges that most users would not want A without also having the functionality provided by B.
-    Package A suggests Package B if B contains files that are related to (and usually enhance) the functionality of A.

If I compare to Bash package has recommendation to Bash-completion
https://packages.debian.org/stretch/bash

Idea was for user to use out-of-box an application (without all build extension but with function extensions of her build extension)
In other way (CI case) admin can add option to not use recommendation :

- in file config : apt::install-recommends "false";
-  on CLI : apt-get install --no-install-recommends...

For the colcon-common-extensions (Virtual Package). I am ok with your vision for one-install-all packages. 

Say to me...

Reply:
Based on you rational (which I agree with) adding `colcon-argcomplete` as a recommends / suggests would not only make sense for `colcon-ros`. It would actually apply to every package which contributes any command line argument to the tool, e.g. `colcon-cmake`. Does that sounds right?

> If I compare to Bash package has recommendation to Bash-completion

The big difference why this is not a perfect comparison is that `bash-completion` doesn't (need to) depend on `bash`. But `colcon-argcomplete` (needs to?) depend on `colcon-core`.

So using the above rational when users use only `colcon-core` we would judge that most of them would want to use completion too, right? Currently it can't recommend / suggest `colcon-argcomplete` though since that package depends on the core itself.

I am not sure if we could / should invert that dependency - remove the dependency on the core from `colcon-argcomplete` and reference it as a recommends / suggests in the core instead :thinking: That would avoid the problem of having each extension package wanting to recommend / suggest the completion...

Reply:
> It would actually apply to every package which contributes any command line argument to the tool, e.g. `colcon-cmake`. Does that sounds right?

From my point of view yes.

> The big difference why this is not a perfect comparison is that bash-completion doesn't (need to) depend on bash

True, it is a bad example. (my bad)

> So using the above rational when users use only `colcon-core` we would judge that most of them would want to use completion too, right?

Right ! if `colcon-core` has any usage of ... but
Actualy no any `colcon-argcomplete` reference on `colcon-core` (I have not analysis how to `colcon-core` use argcomplete... need more investigate.)
_I just make a `colcon_ws/src/colcon-core (master)$ grep -nr colcon_argcomplete ./`._

> But `colcon-argcomplete` (needs to?) depend on `colcon-core`

Right.

> Currently it can't recommend / suggest colcon-argcomplete though since that package depends on the core itself.

Correct, it make a circular reference.

> That would avoid the problem of having each extension package wanting to recommend / suggest the completion...

Of my point of view, it should need !
![dependencies](https://user-images.githubusercontent.com/1913401/44607158-b9248200-a7ef-11e8-9558-ad9f403e708a.png)

Reply:
> currently not possible since it would create a circular dependency

I many case the solution is to split :
- `colcon-core` (or rename to colcon-common) (library)
- `colcon-cli` (app)

![new-approch](https://user-images.githubusercontent.com/1913401/44609470-04db2980-a7f8-11e8-9596-f192f1a30837.png)


In this case
`colcon-cli` can recommend `colcon-argcomplete` and `colcon-argcomplete` depend of `colcon-core`.

> Please describe what the different colors represent (to avoid guessing).

- yellow : language depend
- black : library/app depends
- blue : library/app extra (recommend/suggest)

I make a error between colcon-cmake vs colcon-core (black)

Reply:
> I many case the solution is to split :

That split would prevent to boostrap `colcon` with itself. For a single package needs to contain (enough) to build itself as well as other packages present.

Reply:
My apology, I forgot that need...

Reply:
No problem - many things to consider and juggle at the same time :dizzy_face: